src/Host/Client/Impl/Host/RHost.cs

```diff
@@ -396,16 +396,16 @@ public sealed partial class RHost : IDisposable, IRExpressionEvaluator, IRBlobSe
                     await _transport.CloseAsync();
                     token.Set();
                 } catch (OperationCanceledException) {
+                    token.Set();
                 } catch (MessageTransportException) {
+                    token.Set();
                 } finally {
                     token.Reset();
```

Comment:
This operations are exclusive - each `token` can do either `Set` or `Reset` which immediately releases lock or passes it to the next awaiter. Any subsequent calls of `Set` or `Reset` have no effect.